### PR TITLE
2D top-down camera example

### PR DESCRIPTION
**Pull request by theredfish**
# Objective

This PR addresses the 2D part of #12658. I plan to separate the examples and make one PR per camera example.

## Solution

Added a new top-down example composed of:

- [x] Player keyboard movements
- [x] UI for keyboard instructions
- [x] Colors and bloom effect to see the movement of the player
- [x] Camera smooth movement towards the player (lerp)

## Testing

```bash
cargo run --features="wayland,bevy/dynamic_linking" --example 2d_top_down_camera
```

https://github.com/bevyengine/bevy/assets/10638479/95db0587-e5e0-4f55-be11-97444b795793

### CODE REVIEW

**Comment by theredfish** [comment]
@alice-i-cecile I decided to go with different PRs for the different examples. But if you prefer multiple ones in the same PR let me know.

I kept a simple scope for this one - should be straightforward. I wanted to add more like bounded camera but then I don't think it would serve the purpose of the example. If you think it's valuable here I can add it.

**Comment by alice-i-cecile** [comment]
Multiple small examples is perfect, thank you!

File: examples/camera/2d_top_down_camera.rs
Comment (alice-i-cecile) [inline examples/camera/2d_top_down_camera.rs:32]: We almost certainly want explicit system ordering here.

File: examples/camera/2d_top_down_camera.rs
Comment (alice-i-cecile) [inline examples/camera/2d_top_down_camera.rs:9]: Comments on both of these.

File: examples/camera/2d_top_down_camera.rs
Comment (mgi388) [inline examples/camera/2d_top_down_camera.rs:25]: nit: I think all the examples just inline this rather than making a variable.

File: examples/camera/2d_top_down_camera.rs
Comment (mgi388) [inline examples/camera/2d_top_down_camera.rs:69]: nit: I think all the examples just inline this rather than making a variable.

File: examples/camera/2d_top_down_camera.rs
Comment (mgi388) [inline examples/camera/2d_top_down_camera.rs:34]:
```suggestion
        .add_systems(Startup, (setup_scene, setup_camera).chain())
```

File: examples/camera/2d_top_down_camera.rs
Comment (mgi388) [inline examples/camera/2d_top_down_camera.rs:39]:
```suggestion
fn setup_scene(
```

File: examples/camera/2d_top_down_camera.rs
Comment (mgi388) [inline examples/camera/2d_top_down_camera.rs:68]:
```suggestion
fn setup_camera(mut commands: Commands) {
```

**Review by mgi388 — COMMENTED** [review]
To help other reviewers, you might also consider including a video of the example in the PR description. Check the latest PR template, it has a "Testing" section.

**Comment by theredfish** [comment]
> To help other reviewers, you might also consider including a video of the example in the PR description. Check the latest PR template, it has a "Testing" section.

Thanks for the review! 

Regarding the video, I'll add it but I don't think it's easy for contributors to maintain the video through changes and it can be outdated at some point. Also it adds extra steps, you need tools to capture gif etc... maybe not ideal for everyone.

I'll send code changes soon.

File: examples/camera/2d_top_down_camera.rs
Comment (theredfish) [reply examples/camera/2d_top_down_camera.rs:135]: I think I already normalized the direction with the input handler? `x` and `y` can only be 0.0, -1.0 or 1.0

File: examples/camera/2d_top_down_camera.rs
Comment (theredfish) [reply examples/camera/2d_top_down_camera.rs]: Should be fixed

File: examples/camera/2d_top_down_camera.rs
Comment (theredfish) [reply examples/camera/2d_top_down_camera.rs]: I used the ones from `atmospheric_fog` - Idk if it's the "standard" way but I found similar instructions in other examples.

File: examples/camera/2d_top_down_camera.rs
Comment (mgi388) [reply examples/camera/2d_top_down_camera.rs:160]: See https://forum.unity.com/threads/diagonal-movement-is-quicker.941102/#:~:text=You%20need%20to%20normalize%20the,construct%20from%20the%20input%20axes.&text=As%20you%20can%20see%2C%20when,in%20terms%20of%20a%20square. for example.

The magnitude of the vector exceeds 1 when moving diagonally. I believe you should be able to observe the before and after: before the example moves faster when going diagonally.

Comment (theredfish) [reply examples/camera/2d_top_down_camera.rs]: Thank you! I didn't notice this before. Really interesting and useful to know! Even contributing to examples can help you to learn a lot! I pushed the fix along with a comment - I thought it would be a great thing to add for guidance.

**Comment by theredfish** [comment]
Ok finally found a way to make lighter mp4 records - I attached the record to the PR. If it can provide a good example for other contributions.

File: examples/camera/2d_top_down_camera.rs
Comment (alice-i-cecile) [inline examples/camera/2d_top_down_camera.rs:30]: We should remove this from the example; it's distracting.

File: examples/camera/2d_top_down_camera.rs
Comment (alice-i-cecile) [inline examples/camera/2d_top_down_camera.rs:110]: We should remove these from the example to avoid being distracting. I like keeping the defensive coding though.

**Comment by alice-i-cecile** [comment]
Don't worry about squashing, and thanks for the video! I'm happy with this, barring the complaints about the debug statements. Once those are removed you'll have my approval (and then we just need to find a second reviewer).

**Review by alice-i-cecile — APPROVED** [review]

File: examples/camera/2d_top_down_camera.rs
Comment (rparrett) [inline examples/camera/2d_top_down_camera.rs:32]: These systems are not dependent on each other, so there is no need to `chain` them.

File: examples/camera/2d_top_down_camera.rs
Comment (rparrett) [inline examples/camera/2d_top_down_camera.rs:29]: I would prefer not to change the default clear color for visual consistency with other examples unless there's a good reason.

File: examples/camera/2d_top_down_camera.rs
Comment (rparrett) [inline examples/camera/2d_top_down_camera.rs:130]:
```suggestion
    let mut direction = Vec2::ZERO;
```

File: examples/camera/2d_top_down_camera.rs
Comment (rparrett) [inline examples/camera/2d_top_down_camera.rs:133]: I think these should be all be e.g.

```diff
- direction.y = 1.;
+ direction.y += 1.;
```

So that when left+right are held, the player remains still, rather than right always taking precedence.

File: examples/camera/2d_top_down_camera.rs
Comment (rparrett) [inline examples/camera/2d_top_down_camera.rs:114]: > Here we use the in-game time (in seconds), to get the elapsed time since the previous time update.

The above sentence seems confusing to me. Maybe something like:

> Here we use the in-game time, to get the elapsed time (in seconds) since the previous update.

And there's a typo in the following sentence:

> This avoids jittery when tracking the player.

Seems to be missing the word "movement."

File: examples/camera/2d_top_down_camera.rs
Comment (rparrett) [inline examples/camera/2d_top_down_camera.rs:83]: The z positions of the various objects in this example are:

```
play field: 0
camera: 1
player: 2
```

The z value for the camera doesn't actually matter, but it seems potentially confusing to explicitly set it to this particular value.

I would just use the default transform here.

File: examples/camera/2d_top_down_camera.rs
Comment (BD103) [inline examples/camera/2d_top_down_camera.rs:55]: Could you add a comment here explaining that we use numbers above 1 for bloom?

File: examples/camera/2d_top_down_camera.rs
Comment (theredfish) [reply examples/camera/2d_top_down_camera.rs:133]: Thanks! Indeed!

File: examples/camera/2d_top_down_camera.rs
Comment (theredfish) [reply examples/camera/2d_top_down_camera.rs:29]: It was just to add more contrast for the bloom effect but it's not necessary

**Review by jonathaningram — APPROVED** [review]
Ran it locally, looks good to me. Only nitpick I'd make is that you can move outside of the pretend "bounds" of the example, but I don't think it's a big deal to worry about for this example.

**Comment by theredfish** [comment]
Thanks everyone for the reviews! I Learned a lot through this small example ✌️.

**Pull request merged**
